packages/ws-worker/src/cli.ts

```diff
@@ -0,0 +1,121 @@
+};
+
+function setArg<T>(argValue: T | undefined, envValue: string | undefined, defaultValue: T): T {
+  if (Array.isArray(defaultValue) && envValue !== undefined && argValue == undefined) {
+    return (envValue.split(',') as unknown) as T;
+  }
+
+  if (typeof defaultValue === 'number' && envValue !== undefined && argValue == undefined) {
+    return parseInt(envValue) as unknown as T;
+  }
+
+  return argValue ?? (envValue as unknown as T) ?? defaultValue;
+}
+
+export function parseArgs(argv: string[]): Args {
+  const {
+    WORKER_BACKOFF,
+    WORKER_CAPACITY,
+    WORKER_LIGHTNING_PUBLIC_KEY,
+    WORKER_LIGHTNING_SERVICE_URL,
+    WORKER_LOG_LEVEL,
+    WORKER_MAX_RUN_DURATION_SECONDS,
+    WORKER_MAX_RUN_MEMORY_MB,
+    WORKER_PORT,
+    WORKER_REPO_DIR,
+    WORKER_SECRET,
+    WORKER_STATE_PROPS_TO_REMOVE,
+  } = process.env;
+
+  const parser = yargs(hideBin(argv))
+    .command('server', 'Start a ws-worker server')
+    .option('port', {
+      alias: 'p',
+      description: 'Port to run the server on.',
+      type: 'number',
+    })
+    .option('lightning', {
+      alias: ['l', 'lightning-service-url'],
+      description: 'Base url to Lightning websocket endpoint, eg, ws://localhost:4000/worker. Set to "mock" to use the default mock server.',
+    })
+    .option('repo-dir', {
+      alias: 'd',
+      description: 'Path to the runtime repo (where modules will be installed).',
+    })
+    .option('secret', {
+      alias: 's',
+      description: 'Worker secret.',
+    })
+    .option('lightning-public-key', {
+      description: 'Base64-encoded public key. Used to verify run tokens.',
+    })
+    .option('log', {
+      description: 'Set the log level for stdout (default to info, set to debug for verbose output).',
+    })
+    .option('loop', {
+      description: 'Disable the claims loop',
+      type: 'boolean',
+      default: true,
+    })
+    .option('mock', {
+      description: 'Use a mock runtime engine',
+      type: 'boolean',
+      default: false,
+    })
+    .option('backoff', {
+      description: 'Claim backoff rules: min/max (in seconds).',
+    })
+    .option('capacity', {
+      description: 'max concurrent workers.',
+      type: 'number',
+    })
+    .option('state-props-to-remove', {
+      description: 'A list of properties to remove from the final state returned by a job.',
+      type: 'array',
+    })
+    .option('run-memory', {
+      description: 'Maximum memory allocated to a single run, in mb.',
+      type: 'number',
+    })
+    .option('max-run-duration-seconds', {
+      alias: 't',
+      description: 'Default run timeout for the server, in seconds.',
+      type: 'number',
+    });
+
+  const args = parser.parse() as Args;
+
+  return {
+    ...args,
+    port: setArg(args.port, WORKER_PORT, 2222),
+    lightning: setArg(args.lightning, WORKER_LIGHTNING_SERVICE_URL, 'ws://localhost:4000/worker'),
+    repoDir: setArg(args.repoDir, WORKER_REPO_DIR, undefined),
```

Comment:
I am a bit confused about the approach to handling optional arguments in the `setArg` function.

If we use the optional argument method like `setArg<T>(argValue?: T, envValue?: string, defaultValue?: T)`, we will have to set the return type of the function to `T | undefined`. I am not sure if this is acceptable.

Alternatively, if we use `setArg<T>(defaultValue: T, argValue?: T, envValue?: string)`, where the default value must be passed, we will have to pass `undefined` explicitly since it will no longer be an optional parameter.

I am unsure which approach would be best. Your suggestion would be greatly appreciated.

Reply:
This signature seems correct to me. All parameters are really optional, and the function can quite happily return `undefined`
```
setArg<T>(argValue?: T, envValue?: string, defaultValue?: T)
```
>  we will have to set the return type of the function to `T | undefined`

I'm not sure we do?

It seems to compile without error in the[ Typescript Playground](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwGcQMBBGAcwB4AVAPgApZyA1KCZEAfgC55qAaeCFQA3Vuy68CGGFlTlBoRFGQQM4jjz4BKXtQDcAKENg80+EwLwAvPADeh+E-gAPXmiVyQwR84D0fq68AEQuwYYAvsamqObMJAAyAKoAogD6APIAYmkAGjbwoeGGRKQUjBQEAHQugvHJ6dl5gkXaQA)

In the event that one of the `setArg()` calls returns `undefined` or something, you can always overwrite the generic at call-time. Although I'm not sure why we'd need that?

```
setArg<string>(args.x, process.env.X)
```

If this doesn't compile for you locally then let me know.

Reply:
Thank you for the suggestion. I'll resolve this in the next commit.